### PR TITLE
Remote file upload processor (#33)

### PR DESCRIPTION
**Pull request by pmarjan**
### Description (*)

The implementation of Remote file upload processor (#33).

SourceUploadResponseInterface.php has added getSource() and setSource() which operate with the  object that implements SourceInterface.
This enables to later send back the unique file id through Source's model import_data property.

The source model itself is edited to extend from AbstractExtensibleModel.

### Manual testing scenarios (*)

1. "import_data": "http://some.existing.domain/some-existing-file.csv"
2. "import_data": "http://some.nonexisting.domain/some-non-existing-file.csv"
3. "import_data": "https://some.existing.domain/some-existing-file-wrong-protocol.csv" (e.g. file can be served via http, but not by https)
4. "import_data" not send at all

### CODE REVIEW

**Comment by magento-cicd2** [comment]
[CLA assistant check](https://cla.dev.magento.com/magento/async-import?pullRequest=40) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dev.magento.com/magento/async-import?pullRequest=40) before we can accept your contribution.<br/>

File: app/code/Magento/ImportService/Model/Import/Processor/ExternalFileProcessor.php [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+/**
+ * CSV files processor for asynchronous import
+ */
+class ExternalFileProcessor implements SourceProcessorInterface
+{
+    /**
+     * Magento Media directory
+     */
+    const DIR_MEDIA = BP . '/pub/media/';
```
Comment (nuzil) [inline app/code/Magento/ImportService/Model/Import/Processor/ExternalFileProcessor.php]: I would change location implementation:
1) Make it in "var"
2) Move folder location outside of processor
You can use somethig like that: 
protected $_varDirectory;
\Magento\Framework\Filesystem $filesystem
$this->_varDirectory = $filesystem->getDirectoryWrite(DirectoryList::VAR_DIR);

public function getWorkingDir()
    {
        return $this->_varDirectory->getAbsolutePath('importexport/');
    }

File: app/code/Magento/ImportService/Helper/FileSystem.php [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    public function getWorkingDir() {
+        return $this->varDirectory->getAbsolutePath(self::WORKING_DIR);
+    }
+
+    /**
+     * Get path to temporary directory
+     * @return string 
+     */
+    public function getTemporaryDir() {
+        return $this->mediaDirectory->getAbsolutePath(self::TEMPORARY_DIR);
```
[the remark this answers is not in the document]
Comment (nuzil) [reply app/code/Magento/ImportService/Helper/FileSystem.php]: I would remove "media" from usage fully, cause media is accessible from outside

File: app/code/Magento/ImportService/Helper/FileSystem.php [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+     */
+    public function __construct(
+        Context $context,
+        FrameworkFileSystem $filesystem
+    ) {
+        $this->varDirectory = $filesystem->getDirectoryWrite(DirectoryList::VAR_DIR);
+        $this->mediaDirectory = $filesystem->getDirectoryWrite(DirectoryList::MEDIA);
+
+        if (!is_dir($this->getWorkingDir())) {
+            mkdir($this->getWorkingDir(), 0775, true);
```
Comment (nuzil) [inline app/code/Magento/ImportService/Helper/FileSystem.php]: You dont need that.
On a step when you are copying file with 
lib/internal/Magento/Framework/Filesystem/Directory/Write.php
Its creating directory if needed
see: copyFile() method

Comment (pmarjan) [reply app/code/Magento/ImportService/Helper/FileSystem.php]: Oh, this is good. Thanks for this advice!

File: app/code/Magento/ImportService/Helper/FileSystem.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (pmarjan) [reply app/code/Magento/ImportService/Helper/FileSystem.php]: ExternalFileProcessorTest.php has a method that tests the Processor when existing file is send.
In order to send an existing file, first a temporary directory is created inside the 'media' folder, then a temporary file in it. Media is accessible from outside so an HTTP link can be build in combination with the base url. This link is used by the Test to send to the processor.
Eventually, both the temporary directory and its content are deleted.

But regardless of that, I think I have misused the Magento/Framework/Filesystem class. In the next commit an object of this class is directly used both in the Processor and the Test, so, this Helper is not needed anymore.

Anyway, thank you for pointing out this misplacement.

File: app/code/Magento/ImportService/Model/Import/Processor/ExternalFileProcessor.php [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+        $this->fileSystem = $fileSystem;
+    }
+
+    /**
+     *  {@inheritdoc}
+     */
+    public function processUpload(\Magento\ImportService\Api\Data\SourceInterface $source, \Magento\ImportService\Api\Data\SourceUploadResponseInterface $response)
+    {
+        /** @var string $workingDirectory */
+        $workingDirectory = 'importservice/';
```
Comment (nuzil) [inline app/code/Magento/ImportService/Model/Import/Processor/ExternalFileProcessor.php]: Lets move it in SourceProcessorPool as constant

Comment (pmarjan) [reply app/code/Magento/ImportService/Model/Import/Processor/ExternalFileProcessor.php]: Done. It is moved now.

File: app/code/Magento/ImportService/Model/Source/Validator.php [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+ */
+class Validator
+{
+    /**
+     * The expected mime types for imported files
+     * @var array
+     */
+    const EXPECTED_MIME_TYPES = [
+        'text/plain',
+        'text/csv'
```
Comment (nuzil) [inline app/code/Magento/ImportService/Model/Source/Validator.php]: Can you please implement it in a way as done here: 
https://github.com/magento/async-import/pull/44/files#diff-81a6719121050e174305c001483af652R32

File: app/code/Magento/ImportService/Model/Source/Validator.php [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+    /**
+     * @param \Magento\ImportService\Model\Source $source
+     * @return array|null
+     */
+    public function validateRequest(\Magento\ImportService\Model\Source $source)
+    {
+        $errors = [];
+
+        if (!$this->validateSourceType($source)) {
+            $errors[] = $source::SOURCE_TYPE . ' cannot be empty';
```
Comment (nuzil) [inline app/code/Magento/ImportService/Model/Source/Validator.php]: Can you also make it translatable
_(' %')

File: app/code/Magento/ImportService/Model/Source/Validator.php (hunk 5 above)
Comment (pmarjan) [reply app/code/Magento/ImportService/Model/Source/Validator.php]: Done. Moved to di.xml

File: app/code/Magento/ImportService/Model/Source/Validator.php (hunk 6 above)
Comment (pmarjan) [reply app/code/Magento/ImportService/Model/Source/Validator.php]: Done. I also added i18n/en_US.csv file.

File: app/code/Magento/ImportService/Model/Import/Processor/ExternalFileProcessor.php [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+
+/**
+ * CSV files processor for asynchronous import
+ */
+class ExternalFileProcessor implements SourceProcessorInterface
+{
+    /**
+     * @var \Magento\Framework\Filesystem
+     */
+    protected $fileSystem;
```
Comment (vkublytskyi) [inline app/code/Magento/ImportService/Model/Import/Processor/ExternalFileProcessor.php]: Please use only `private` and `public` visibility for properties and methods. `protected` may be used only in a case when you must inherit from a legacy class.

Comment (pmarjan) [reply app/code/Magento/ImportService/Model/Import/Processor/ExternalFileProcessor.php]: Done. `protected` is changed and now for class properties `private` is used.

**Pull request merged**